### PR TITLE
Align Pi menu for nodes

### PR DESCRIPTION
**Pull request by 3DSinghVFX**
I have added an *Align Pi* menu that allows to align the nodes w.r.t active node. It has eight buttons, two for dependent nodes and dependencies nodes, the other four for selection nodes:
Shortcut Key: F
- Dependencies - That aligns the all dependencies nodes w.r.t active node.
- Dependent  - That aligns the all dependent nodes w.r.t active node.
- Selection (Top)  - That stacks up the all selected nodes w.r.t active node.
- Selection (Down)  - That stacks down the all selected nodes w.r.t active node.
- Selection (Top Right)  - That aligns the headers of all selected nodes w.r.t active node.
- Selection (Top Left)  - That aligns the headers of all selected nodes w.r.t active node.
- Selection (Bottom Right)  - That aligns the sides of all selected nodes w.r.t active node to the right side.
- Selection (Bottom Left)  - That aligns the sides of all selected nodes w.r.t active node to the left side.

Peek 2020-08-17 13-03

Screenshot from 2020-08-17 13-04-20

### CODE REVIEW

**Comment by 3DSinghVFX** [comment]
> Most of the operators doesn't seem to be working correctly for me.
> test

Omar, that is because of the selection order. I have noticed that the E-Pie menu gives the wrong selection order.

File: animation_nodes/keymap.py [hunk 1]
```diff
@@ -22,6 +22,10 @@ def register():
     kmi = km.keymap_items.new("wm.call_menu_pie", type = "E", value = "PRESS")
     kmi.properties.name = "AN_MT_selection_pie"
 
+    # Selection Pie Menu
+    kmi = km.keymap_items.new("wm.call_menu_pie", type = "F", value = "PRESS")
```
[the remark this answers is not in the document]
Comment (3DSinghVFX) [reply animation_nodes/keymap.py:26]: Should we use `Ctrl+E`?

File: animation_nodes/keymap.py [hunk 2]
```diff
@@ -22,6 +22,10 @@ def register():
     kmi = km.keymap_items.new("wm.call_menu_pie", type = "E", value = "PRESS")
     kmi.properties.name = "AN_MT_selection_pie"
 
+    # Selection Pie Menu
```
[the remark this answers is not in the document]
Comment (3DSinghVFX) [reply animation_nodes/keymap.py:25]: Changed.

**Comment by OmarEmaraDev** [comment]
Shouldn't the operator be independent of the selection order?

File: animation_nodes/keymap.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (OmarEmaraDev) [reply animation_nodes/keymap.py:26]: Ok. Just make sure it is not used first.

[the remark this answers is not in the document]
Comment (3DSinghVFX) [reply animation_nodes/keymap.py:26]: Okay.

**Comment by 3DSinghVFX** [comment]
> Shouldn't the operator be independent of the selection order?

Actually, the *Dependent* and *Dependencies* operators are independent of the selection order. However, for non-connected nodes or are the nodes that someone wants to align vertical need specific order that is why selection order is required.

**Comment by 3DSinghVFX** [comment]
@OmarEmaraDev Is there a way to make the `Offset` parameter available in the Preferences?

**Comment by OmarEmaraDev** [comment]
The dependency-based alignment operators don't work with branched nodes either.
test

**Comment by 3DSinghVFX** [comment]
> The dependency-based alignment operators don't work with branched nodes either.
> test

Is there a way to find the branched nodes?

**Comment by OmarEmaraDev** [comment]
I am not sure, but you can certainly write your own methods.
I think the operators should be completely automatic and general enough to work in all cases with no user intervention, other than that, I don't really consider this useful. Maybe something like JuhaW's NodeArrange add-on.

**Comment by 3DSinghVFX** [comment]
> I am not sure, but you can certainly write your own methods.

Omar, One way is, only align the nodes up to the branch point. To align all branches seems difficult to me.

> I think the operators should be completely automatic and general enough to work in all cases with no user intervention, other than that, I don't really consider this useful. Maybe something like JuhaW's NodeArrange add-on.
 
Okay.

**Comment by 3DSinghVFX** [comment]
@OmarEmaraDev I have added support for branched nodes for *Dependencies* and *Dependent*.

**Comment by 3DSinghVFX** [comment]
> Multiple outputs do not work.

I'll try to fix that.

**Comment by 3DSinghVFX** [comment]
@OmarEmaraDev  I have fixed the *Multiple Outputs*,
Peek 2020-08-18 17-07

**Comment by Gerstmann-Bradley** [comment]
Hi, I am just curious: what If I have rerouting within this node tree, how will they be effected?

**Comment by 3DSinghVFX** [comment]
> Hi, I am just curious: what If I have rerouting within this node tree, how will they be effected?

Nice point. I'll try include that in alignment.

**Comment by 3DSinghVFX** [comment]
> Hi, I am just curious: what If I have rerouting within this node tree, how will they be effected?

@Gerstmann-Bradley Now, it takes care of *reroutes*:
Peek 2020-08-18 17-47

**Comment by JacquesLucke** [comment]
I'm honestly not sure if such an operator should be part of Animation Nodes. Getting it to function well enough is not a simple task. Also, this functionality is useful outside of Animation Nodes. I think it would be better to outsource this functionality into a separate addon, that can be installed without AN and maybe even become part of Blender at some point.

**Comment by 3DSinghVFX** [comment]
> I'm honestly not sure if such an operator should be part of Animation Nodes. Getting it to function well enough is not a simple task. Also, this functionality is useful outside of Animation Nodes. I think it would be better to outsource this functionality into a separate addon, that can be installed without AN and maybe even become part of Blender at some point.

@JacquesLucke Thanks for your suggestions. I'll try to make this functionality into a separate addon.

**Comment by OmarEmaraDev** [comment]
I think we all agree now that this is better suited as a separate add-on, so I will close this for now.

**Comment by 3DSinghVFX** [comment]
> I think we all agree now that this is better suited as a separate add-on, so I will close this for now.

@JacquesLucke  and @OmarEmaraDev thanks. The **Align Nodes** addon is now available https://github.com/3DSinghVFX/align_nodes

**Pull request closed without merging**